cocos2d/Support/CCFileUtilsV2.m

```diff
@@ -0,0 +1,373 @@
+}
+
+- (CCFile *)findFileInAllSearchPaths:(NSString *)filename metaData:(CCFileResolvedMetaData *)metaData options:(NSDictionary *)options
+{
+    for (NSString *searchPath in _searchPaths)
+    {
+        NSString *resolvedFilename = filename;
+
+        if (metaData)
+        {
+            resolvedFilename = metaData.filename;
+        }
+
+        CCFile *aFile = [self findFilename:resolvedFilename inSearchPath:searchPath options:options];
+
+        if (aFile)
+        {
+            if (metaData)
+            {
+                aFile.useUIScale = metaData.useUIScale;
+            }
+
+            _cache[filename] = aFile;
+            return aFile;
+        }
+    }
+    return nil;
+}
+
+- (CCFileResolvedMetaData *)resolvedMetaDataForFilename:(NSString *)filename
+{
+    if (!_database)
+    {
+        return nil;
+    }
+
+    for (NSString *searchPath in _searchPaths)
+    {
+        CCFileMetaData *metaData = [_database metaDataForFileNamed:filename inSearchPath:searchPath];
+
+        if (!metaData)
+        {
+            continue;
+        }
+
+        TraceLog(@"* Database entry for search path \"%@\": %@", searchPath, metaData);
+        return [self resolveMetaData:metaData];
+    }
+
+    return nil;
+}
+
+- (CCFileResolvedMetaData *)resolveMetaData:(CCFileMetaData *)metaData
+{
+    CCFileResolvedMetaData *fileUtilsV2ResolvedMetaData = [[CCFileResolvedMetaData alloc] init];
+
+    NSString *localizedFileName = [self localizedFilenameWithMetaData:metaData];
+
+    fileUtilsV2ResolvedMetaData.filename = localizedFileName
+        ? localizedFileName
+        : metaData.filename;
+
+    fileUtilsV2ResolvedMetaData.useUIScale = metaData.useUIScale;
+
+    return fileUtilsV2ResolvedMetaData;
+}
+
+- (NSString *)localizedFilenameWithMetaData:(CCFileMetaData *)metaData
+{
+    if (!metaData.localizations)
+    {
+        return nil;
+    }
+
+    for (NSString *languageID in [NSLocale preferredLanguages])
+    {
+        NSString *filenameForLanguageID = metaData.localizations[languageID];
+        if (filenameForLanguageID)
+        {
+            TraceLog(@"* Localization for languageID \"%@\" found: \"%@\"", languageID, filenameForLanguageID);
+            return filenameForLanguageID;
+        }
+    }
+
+    return nil;
+}
+
+- (CCFile *)findFilename:(NSString *)filename inSearchPath:(NSString *)searchPath options:(NSDictionary *)options
+{
+    NSArray *searchVariants = [self contentScaleVariantsWithFilename:filename options:options];
+
+    TraceLog(@"* Searching for content scale variants: %@", searchVariants);
+
+    for (CCFileUtilsV2SearchData *fileLocatorSearchData in searchVariants)
+    {
+        NSURL *fileURL = [NSURL fileURLWithPath:[searchPath stringByAppendingPathComponent:fileLocatorSearchData.filename]];
+
+        BOOL isDirectory = NO;
+        NSFileManager *fileManager = [NSFileManager defaultManager];
+        if ([fileManager fileExistsAtPath:fileURL.path isDirectory:&isDirectory] && !isDirectory)
```

Comment:
Need to add a check for a local URL.

Until now, I hadn't considered that you can't just query for file existence on remote servers. We would have to make the request and check for errors. This could make loading files from HTTP *ridiculously* slow if it has to make a few failing attempts before it finds the right file.

My main reason for wanting to support remote servers was to support reloading assets at runtime on a device. The performance wouldn't really be a problem there. I guess people would figure out the performance implications of remote file pretty quickly.